### PR TITLE
client/web: skip check mode for non-tailscale.com control servers

### PR DESCRIPTION
**Pull request by willnorris**
Only enforce check mode if the control server URL ends in ".tailscale.com".  This allows the web client to be used with headscale (or other) control servers while we work with the project to add check mode support (tracked in juanfont/headscale#1623).

Updates #10261

I don't *love* all of the string pointers in the tests, but it does seem to work fine and is consistent with how we do this in some other tests.  It does mean that we have to be careful not to run these test cases in parallel, but that's probably fine.

### CODE REVIEW

File: client/web/auth.go [hunk 1]
```diff
@@ -160,14 +158,44 @@ func (s *Server) newSession(ctx context.Context, src *apitype.WhoIsResponse) (*b
...
 	return session, nil
 }
 
+// controlSupportsCheckMode returns whether the current control server supports web client check mode.
+// We assume that only "tailscale.com" control servers support check mode.
+// This allows the web client to be used with non-standard control servers.
+// If an error occurs getting the control URL, this method returns true to fail closed.
+//
+// TODO(juanfont/headscale#1623): adjust or remove this when headscale supports check mode.
+func (s *Server) controlSupportsCheckMode(ctx context.Context) bool {
```
Comment (soniaappasamy) [inline client/web/auth.go:187]: The other thing semi-related to this is right now we have some "login.tailscale.com" admin panel hard-coded links on the frontend. I think maybe we just hide those for any non-tailscale control servers. We can address that in a different change, just a thought.

File: client/web/auth.go [hunk 2]
```diff
@@ -160,14 +158,44 @@ func (s *Server) newSession(ctx context.Context, src *apitype.WhoIsResponse) (*b
...
+	} else {
+		// control does not support check mode, so there is no additional auth we can do.
+		session.Authenticated = true
+	}
+
 	s.browserSessions.Store(sid, session)
 	return session, nil
 }
 
+// controlSupportsCheckMode returns whether the current control server supports web client check mode.
```
Comment (soniaappasamy) [inline client/web/auth.go:181]:
```suggestion
// controlSupportsCheckMode returns whether the current control server supports web client check mode, to verify a user's identity.
```

**Review by soniaappasamy — APPROVED** [review]

File: client/web/auth.go [hunk 3]
```diff
@@ -160,14 +158,44 @@ func (s *Server) newSession(ctx context.Context, src *apitype.WhoIsResponse) (*b
...
+// We assume that only "tailscale.com" control servers support check mode.
+// This allows the web client to be used with non-standard control servers.
+// If an error occurs getting the control URL, this method returns true to fail closed.
+//
+// TODO(juanfont/headscale#1623): adjust or remove this when headscale supports check mode.
+func (s *Server) controlSupportsCheckMode(ctx context.Context) bool {
+	prefs, err := s.lc.GetPrefs(ctx)
+	if err != nil {
+		return true
+	}
```
[the remark this answers is not in the document]
Comment (willnorris) [reply client/web/auth.go:191]: Yeah, I had that thought as well... my initial implementation was doing it on every request, and I switched to this much simpler approach.  Calling it just on new session creation should be fine, and as you say, a little safer too.

[the remark this answers is not in the document]
Comment (soniaappasamy) [reply client/web/auth.go:191]: Yeah I really like your approach of still creating a session either way. Simplifies things a ton

**Pull request merged**